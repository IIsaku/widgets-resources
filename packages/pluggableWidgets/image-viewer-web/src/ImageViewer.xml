<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.custom.imageviewer.ImageViewer" needsEntityContext="false" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd" pluginWidget="true">
    <name>Image viewer</name>
    <description>Display an image and enlarge it on click</description>
    <icon>iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAMAAACdt4HsAAABnlBMVEUAAAD///+A//9V//+Av/9mzP9gv/9Vxv9Ns/9duf9Vv/9Osf9Jtv9Vu/9Qv/9LtP9RvP9Jtv9Osf9Ktf9Ms/9Puf9Ns/9Ktf9Ms/hKtfhOsfhNs/lLtPlLtPlKtflIsflMs/pJsfpNs/pIsvpKsPpLtPtKsftJsvtJsvdJs/hLtPhKsfhJsvhXV1dLsflKsvlKsvlKsvlKsvpKsvdKs/dJsfdIsvhKsPhJsfhKsPhJsvhKsfhIsPlKsflJsPlKsflIsflJsvdKsvdJsfhJsPhJsfhKsvhIsvhJsfhIsPhKsfhJsvlJsfdJsfdIsPdKsfdJsfdJsfdIsfdIsfhIsfhIsfhJsfhJsfhIsPhJsPhIsfhJsfdJsfdJsfdJsfdIsPdJsfdJsPhIsfhJsfhIsfhJsfhIsfhJsfhJsPhJsfhJsfhIsfhIsPhJsfdIsPdJsPdVVVVJsfdJsfdJsfdIsPdJsfdIsPdIsPdJsfhJsPhIsPhJsfhIsPhIsfhJsfhJsPhIsfhJsfhJsPhIsPhIsfhJsPhJsPhIsfhIsPdVVVU8lPiOAAAAiHRSTlMAAQIDBAUICQoLDA0ODxARExUXGBsdHh8lJicoKSwtLi8xMjU3PT4/QkZHSElJUlNWWV1gZGVqa2xucHV4eXt8f4SHiYuMjpGTm5yeoKGio6SnqaqtsbO2uLm+wMHEyMnLzM3O0NHU1dba3d7f4eLj4+Tn6Onr7O3u7/Dy8/T19vf4+fr7/P3+LOCsiQAAAeRJREFUeAHt1/lTEmEYwPHHhYyiLEjtPtLu0siOsLL7MLtNOjRLs8MyU7Q7K4hyv/7Xve/sLrAzArPvMFPT7Oen9wf4MsDz7OyKI9QxvWBgukM8Ru9XBfEsGKpvQAL6/wORY48+5ieu7zIN7HiG4+5yo0B7Hs/oMoNA4i0lN4IHkreA+Ssbmva/BOztAQP7nqL1iBJ7BfQGCxy30WYt0Q4C44ECbb9Qih+7BpgLFLgN2BOF0WbTQBboFEsUs6/wDXBnx+xHfAKcXuKcV5j8jZdwmQ7SqhlcpqO8e5Yi/zJt6ZvMTfXvrBWQRO+YOws/feucLqDZFwwvKJ02rrRRIPkBz/dmk8AAkDsU3/sJuGkQSKGcVIejgN1eM9DYfX/yx+uH3S3iSLwHHkfUyXoODNcKdGVx/L62VrR+IL9JtD0oqaoB6yIlubS7UpwSRwZ401gl0HAHn0xs9TtgJCKOdQWgp0rgPEr+3PrYxhMv0IYH9FBtFs9l4GuyYqALJeu83Dr8GdcZKWrSI9FXKdAyB0y1iqt1BI2xqJSkgfltFQL39Ki1SVF8CKWwVcpEx4HBxQMplCNSZukgcFZ8DqAsGojPAEMNUm5l5svVqPg9cAP/2D1SeJcWBsLA398F84euOj32hf4AR3v7uNFLZj0AAAAASUVORK5CYII=</icon>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Data source">
                <property key="datasource" type="enumeration" defaultValue="image">
                    <caption>Data source</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="image">Image</enumerationValue>
                        <enumerationValue key="imageUrl">Image URL</enumerationValue>
                        <enumerationValue key="icon">Icon</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="imageObject" type="image" required="false">
                    <caption>Image</caption>
                    <description>The image from the modeler</description>
                </property>
                <property key="imageUrl" type="textTemplate" required="false">
                    <caption>Image URL</caption>
                    <description>The URL of the external image</description>
                </property>
                <property key="imageIcon" type="icon" required="false">
                    <caption>Icon</caption>
                    <description>The icon image</description>
                </property>
            </propertyGroup>
            <propertyGroup caption="Dimensions">
                <property key="widthUnit" type="enumeration" defaultValue="auto">
                    <caption>Width unit</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="auto">Auto</enumerationValue>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                        <enumerationValue key="percentage">Percentage</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="width" type="integer" defaultValue="100">
                    <caption>Width</caption>
                    <description/>
                </property>
                <property key="heightUnit" type="enumeration" defaultValue="auto">
                    <caption>Height unit</caption>
                    <description>Auto will keep the aspect ratio of the image</description>
                    <enumerationValues>
                        <enumerationValue key="auto">Auto</enumerationValue>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="height" type="integer" defaultValue="100">
                    <caption>Height</caption>
                    <description/>
                </property>
                <property key="iconSize" type="integer" defaultValue="14">
                    <caption>Icon size</caption>
                    <description>The size of the icon in pixels</description> 
                </property>
                <property key="responsive" type="boolean" defaultValue="true">
                    <caption>Responsive</caption>
                    <description>Image will never get larger than its original size. It can become smaller</description>
                </property>
            </propertyGroup>
            <!-- TODO: Events -->
        </propertyGroup>
    </properties>
</widget>
