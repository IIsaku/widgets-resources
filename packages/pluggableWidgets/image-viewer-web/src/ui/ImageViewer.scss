.mx-image-viewer {
    height: 100%;
    width: 100%;
}

.mx-image-viewer-responsive img {
    max-width: 100%;
    max-height: 100%;
}
