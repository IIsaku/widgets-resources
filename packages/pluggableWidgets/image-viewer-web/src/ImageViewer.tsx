import { ValueStatus } from "mendix";
import { createElement, FunctionComponent } from "react";
import { ImageViewerContainerProps } from "../typings/ImageViewerProps";
import { ImageViewer as ImageViewerComponent } from "./components/ImageViewer";

type ImageProps = {
    type: "image" | "icon";
    image: string | undefined;
};

function getImageProps({ datasource, imageIcon, imageObject, imageUrl }: ImageViewerContainerProps): ImageProps {
    const fallback: ImageProps = {
        type: "image",
        image: undefined
    };
    switch (datasource) {
        case "image":
            return {
                type: "image",
                image: imageObject?.status === ValueStatus.Available ? imageObject.value.uri : undefined
            };
        case "imageUrl":
            return {
                type: "image",
                image: imageUrl?.status === ValueStatus.Available ? imageUrl.value : undefined
            };
        case "icon": {
            if (imageIcon?.status === ValueStatus.Available) {
                if (imageIcon.value?.type === "glyph") {
                    return {
                        type: "icon",
                        image: imageIcon.value.iconClass
                    };
                }
                if (imageIcon.value?.type === "image") {
                    return {
                        type: "image",
                        image: imageIcon.value.iconUrl
                    };
                }
            }
            return fallback;
        }
        default:
            return fallback;
    }
}

export const ImageViewer: FunctionComponent<ImageViewerContainerProps> = props => {
    const { type, image } = getImageProps(props);
    return (
        <ImageViewerComponent.Wrapper
            className={props.class}
            responsive={props.responsive}
            hasImage={image !== undefined && image.length > 0}
        >
            {type === "image" ? (
                <ImageViewerComponent.Image
                    image={image}
                    height={props.height}
                    heightUnit={props.heightUnit}
                    width={props.width}
                    widthUnit={props.widthUnit}
                    style={props.style}
                />
            ) : (
                <ImageViewerComponent.Glyphicon icon={image} size={props.iconSize} style={props.style} />
            )}
        </ImageViewerComponent.Wrapper>
    );
};
