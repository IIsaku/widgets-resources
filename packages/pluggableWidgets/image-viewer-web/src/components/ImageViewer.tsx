import { createElement, CSSProperties, ReactElement } from "react";
import classNames from "classnames";
import { HeightUnitEnum, WidthUnitEnum } from "../../typings/ImageViewerProps";

import "../ui/ImageViewer.scss";

function getStyle(value: string | number, type: WidthUnitEnum | HeightUnitEnum): number | string {
    // when type is auto default browser styles applies
    if (type === "pixels") {
        return value;
    } else if (type === "percentage") {
        return value + "%";
    }

    return "";
}
export interface ImageViewerWrapperProps {
    className?: string;
    responsive: boolean;
    hasImage: boolean;
    children: ReactElement<ImageViewerGlyphicon | ImageViewerImage>;
}

interface ImageViewerContent {
    style?: CSSProperties;
}

function Wrapper(props: ImageViewerWrapperProps): ReactElement {
    return (
        <div
            className={classNames(
                "mx-image-viewer",
                { "mx-image-viewer-responsive": props.responsive },
                props.className,
                { hidden: !props.hasImage }
            )}
        >
            {props.children}
        </div>
    );
}

export interface ImageViewerGlyphicon extends ImageViewerContent {
    icon: string | undefined;
    size: number;
}

function Glyphicon(props: ImageViewerGlyphicon): ReactElement {
    return (
        <span className={classNames("glyphicon", props.icon)} style={{ ...props.style, fontSize: `${props.size}px` }} />
    );
}

export interface ImageViewerImage extends ImageViewerContent {
    image: string | undefined;
    height: number;
    heightUnit: HeightUnitEnum;
    width: number;
    widthUnit: WidthUnitEnum;
}

function Image(props: ImageViewerImage): ReactElement {
    return (
        <img
            src={props.image}
            style={{
                ...props.style,
                height: getStyle(props.height, props.heightUnit),
                width: getStyle(props.width, props.widthUnit)
            }}
        />
    );
}

export const ImageViewer = {
    Wrapper,
    Glyphicon,
    Image
};
