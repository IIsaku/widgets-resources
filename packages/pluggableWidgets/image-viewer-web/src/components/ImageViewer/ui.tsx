import { createElement, CSSProperties, HTMLAttributes, ReactElement, ReactEventHandler } from "react";
import classNames from "classnames";
import { HeightUnitEnum, WidthUnitEnum } from "../../../typings/ImageViewerProps";
import { LightboxProps } from "../Lightbox";

import "../../ui/ImageViewer.scss";

function getStyle(value: string | number, type: WidthUnitEnum | HeightUnitEnum): number | string {
    // when type is auto default browser styles applies
    if (type === "pixels") {
        return value;
    } else if (type === "percentage") {
        return value + "%";
    }

    return "";
}
export interface ImageViewerWrapperProps {
    className?: string;
    responsive: boolean;
    hasImage: boolean;
    children:
        | ReactElement<ImageViewerGlyphicon | ImageViewerImage>
        | [ReactElement<ImageViewerGlyphicon | ImageViewerImage>, ReactElement<LightboxProps> | false];
}

export interface ImageViewerContentProps {
    style?: CSSProperties;
    onClick?: ReactEventHandler<HTMLElement>;
    altText?: string;
}

function Wrapper(props: ImageViewerWrapperProps): ReactElement {
    return (
        <div
            className={classNames(
                "mx-image-viewer",
                { "mx-image-viewer-responsive": props.responsive },
                props.className,
                { hidden: !props.hasImage }
            )}
        >
            {props.children}
        </div>
    );
}

export interface ImageViewerGlyphicon extends ImageViewerContentProps {
    icon: string | undefined;
    size: number;
}

function Glyphicon(props: ImageViewerGlyphicon): ReactElement {
    const accessibilityProps = props.altText
        ? {
              "aria-label": props.altText,
              role: "img"
          }
        : {};

    const onClickProps = getImageContentOnClickProps(props.onClick);

    return (
        <span
            className={classNames("glyphicon", props.icon)}
            style={{ ...props.style, fontSize: `${props.size}px` }}
            {...accessibilityProps}
            {...onClickProps}
        />
    );
}

export interface ImageViewerImage extends ImageViewerContentProps {
    image: string | undefined;
    height: number;
    heightUnit: HeightUnitEnum;
    width: number;
    widthUnit: WidthUnitEnum;
}

function Image(props: ImageViewerImage): ReactElement {
    const onClickProps = getImageContentOnClickProps(props.onClick);
    return (
        <img
            src={props.image}
            style={{
                ...props.style,
                height: getStyle(props.height, props.heightUnit),
                width: getStyle(props.width, props.widthUnit)
            }}
            alt={props.altText}
            {...onClickProps}
        />
    );
}

function getImageContentOnClickProps(onClick: ImageViewerContentProps["onClick"]): HTMLAttributes<HTMLElement> {
    if (!onClick) {
        return {};
    }
    return {
        onClick,
        role: "button",
        tabIndex: 0,
        onKeyDown: event => {
            if (event.key === "Enter" || event.key === " ") {
                onClick(event);
            }
        }
    };
}

export const ImageViewerUi = {
    Wrapper,
    Glyphicon,
    Image
};
