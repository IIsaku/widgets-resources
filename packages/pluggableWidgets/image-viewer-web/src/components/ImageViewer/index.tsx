import { cloneElement, createElement, CSSProperties, FunctionComponent, useCallback } from "react";
import { HeightUnitEnum, WidthUnitEnum, OnClickTypeEnum, DisplayAsEnum } from "../../../typings/ImageViewerProps";
import { useLightboxState } from "../../utils/lightboxState";
import { ImageViewerUi, ImageViewerContentProps } from "./ui";
import { Lightbox, LightboxProps } from "../Lightbox";

import "../../ui/ImageViewer.scss";

export type ImageViewerImageProps = {
    type: "image" | "icon";
    image: string | undefined;
};

export interface ImageViewerProps extends ImageViewerImageProps {
    class: string;
    style?: CSSProperties;
    widthUnit: WidthUnitEnum;
    width: number;
    heightUnit: HeightUnitEnum;
    height: number;
    iconSize: number;
    responsive: boolean;
    onClickType: OnClickTypeEnum;
    onClick?: () => void;
    altText?: string;
    displayAs: DisplayAsEnum;
    previewMode?: boolean;
}

function processImageLink(imageLink: string | undefined, displayAs: DisplayAsEnum): string | undefined {
    if (!imageLink || displayAs === "fullImage") {
        return imageLink;
    }
    const url = new URL(imageLink);
    url.searchParams.append("thumb", "true");
    return url.href;
}

export const ImageViewer: FunctionComponent<ImageViewerProps> = ({
    class: className,
    style,
    widthUnit,
    width,
    heightUnit,
    height,
    iconSize,
    responsive,
    onClickType,
    onClick,
    type,
    image,
    altText,
    displayAs,
    previewMode
}) => {
    const { lightboxIsOpen, openLightbox, closeLightbox } = useLightboxState();

    const onCloseLightbox = useCallback<LightboxProps["onClose"]>(
        event => {
            event?.stopPropagation();
            closeLightbox();
        },
        [closeLightbox]
    );

    const onImageClick = useCallback<Exclude<ImageViewerContentProps["onClick"], undefined>>(
        event => {
            event.stopPropagation();
            if (onClickType === "action") {
                onClick?.();
            } else if (onClickType === "enlarge") {
                openLightbox();
            }
        },
        [onClick, onClickType, openLightbox]
    );

    const hasClickHandler = (onClickType === "action" && onClick) || onClickType === "enlarge";
    const sharedContentProps: ImageViewerContentProps = {
        style,
        onClick: hasClickHandler ? onImageClick : undefined,
        altText
    };

    const content =
        type === "image" ? (
            <ImageViewerUi.Image
                image={processImageLink(image, displayAs)}
                height={height}
                heightUnit={heightUnit}
                width={width}
                widthUnit={widthUnit}
                {...sharedContentProps}
            />
        ) : (
            <ImageViewerUi.Glyphicon icon={image} size={iconSize} {...sharedContentProps} />
        );

    return (
        <ImageViewerUi.Wrapper
            className={className}
            responsive={responsive}
            hasImage={image !== undefined && image.length > 0}
        >
            {content}
            {!previewMode && lightboxIsOpen && (
                <Lightbox isOpen={lightboxIsOpen} onClose={onCloseLightbox}>
                    {type === "image" ? cloneElement(content, { image, onClick: undefined }) : content}
                </Lightbox>
            )}
        </ImageViewerUi.Wrapper>
    );
};
