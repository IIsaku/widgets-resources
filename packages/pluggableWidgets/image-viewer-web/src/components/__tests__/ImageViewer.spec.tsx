import { createElement } from "react";
import { render } from "enzyme";

import { ImageViewer, ImageViewerImage, ImageViewerGlyphicon } from "../ImageViewer";

describe("ImageViewer", () => {
    const imageProps: ImageViewerImage = {
        image: "https://pbs.twimg.com/profile_images/1905729715/llamas_1_.jpg",
        height: 300,
        heightUnit: "pixels",
        width: 300,
        widthUnit: "pixels"
    };
    const glyphiconProps: ImageViewerGlyphicon = {
        icon: "glyphicon-asterisk",
        size: 20
    };

    it("renders the structure with an image", () => {
        expect(
            render(
                <ImageViewer.Wrapper responsive hasImage>
                    <ImageViewer.Image {...imageProps} />
                </ImageViewer.Wrapper>
            )
        ).toMatchSnapshot();
    });

    it("renders the structure with an image and percentage dimensions", () => {
        expect(
            render(
                <ImageViewer.Wrapper responsive hasImage>
                    <ImageViewer.Image
                        image={imageProps.image}
                        height={100}
                        width={100}
                        heightUnit="auto"
                        widthUnit="percentage"
                    />
                </ImageViewer.Wrapper>
            )
        ).toMatchSnapshot();
    });

    it("renders the structure with an icon", () => {
        expect(
            render(
                <ImageViewer.Wrapper responsive hasImage>
                    <ImageViewer.Glyphicon {...glyphiconProps} />
                </ImageViewer.Wrapper>
            )
        ).toMatchSnapshot();
    });
});
