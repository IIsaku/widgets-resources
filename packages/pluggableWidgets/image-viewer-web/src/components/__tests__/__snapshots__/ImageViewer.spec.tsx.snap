// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageViewer renders the structure with an icon 1`] = `
<div
  class="mx-image-viewer mx-image-viewer-responsive"
>
  <span
    class="glyphicon glyphicon-asterisk"
    style="font-size:20px"
  />
</div>
`;

exports[`ImageViewer renders the structure with an image 1`] = `
<div
  class="mx-image-viewer mx-image-viewer-responsive"
>
  <img
    src="https://pbs.twimg.com/profile_images/1905729715/llamas_1_.jpg"
    style="height:300px;width:300px"
  />
</div>
`;

exports[`ImageViewer renders the structure with an image and percentage dimensions 1`] = `
<div
  class="mx-image-viewer mx-image-viewer-responsive"
>
  <img
    src="https://pbs.twimg.com/profile_images/1905729715/llamas_1_.jpg"
    style="height:;width:100%"
  />
</div>
`;
